cmd/ncproxy/run.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/go-winio/pkg/etwlogrus"
+	"github.com/Microsoft/hcsshim/cmd/ncproxy/nodenetsvc"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/plugin/ocgrpc"
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+)
+
+// Run ncproxy
+func run() error {
+	flag.Parse()
+	// Stop if we are registering or unregistering against Windows SCM.
+	stop, err := registerUnregisterService()
+	if err != nil {
+		return err
+	}
+	if stop {
+		return nil
+	}
+
+	// Provider ID: cf9f01fe-87b3-568d-ecef-9f54b7c5ff70
+	// Hook isn't closed explicitly, as it will exist until process exit.
+	if hook, err := etwlogrus.NewHook("Microsoft.Virtualization.NCProxy"); err == nil {
+		logrus.AddHook(hook)
+	} else {
+		logrus.Error(err)
+	}
+
+	// Register our OpenCensus logrus exporter
+	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
+	trace.RegisterExporter(&oc.LogrusExporter{})
+
+	ctx := context.Background()
+	conf, err := loadConfig(*configPath)
+	if err != nil {
+		return errors.Wrap(err, "failed getting configuration file")
+	}
+
+	if conf.GRPCAddr == "" {
+		return errors.New("missing GRPC endpoint in config")
+	}
+
+	if conf.TTRPCAddr == "" {
+		return errors.New("missing TTRPC endpoint in config")
+	}
+
+	// If there's a node network service in the config, assign this to our global client.
+	if conf.NodeNetSvcAddr != "" {
+		log.G(ctx).Infof("Connecting to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+
+		dialCtx := ctx
+		opts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithStatsHandler(&ocgrpc.ClientHandler{})}
+		if conf.Timeout > 0 {
+			var cancel context.CancelFunc
+			dialCtx, cancel = context.WithTimeout(ctx, time.Duration(conf.Timeout)*time.Second)
+			defer cancel()
+			opts = append(opts, grpc.WithBlock())
+		}
+		client, err := grpc.DialContext(dialCtx, conf.NodeNetSvcAddr, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to connect to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+		}
+
+		log.G(ctx).Infof("Successfully connected to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+
+		netSvcClient := nodenetsvc.NewNodeNetworkServiceClient(client)
+		nodeNetSvcClient = &nodeNetSvcConn{
+			addr:     conf.NodeNetSvcAddr,
+			client:   netSvcClient,
+			grpcConn: client,
+		}
+	}
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"TTRPCAddr":      conf.TTRPCAddr,
+		"NodeNetSvcAddr": conf.NodeNetSvcAddr,
+		"GRPCAddr":       conf.GRPCAddr,
+		"Timeout":        conf.Timeout,
+	}).Info("starting ncproxy")
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT)
```

Comment:
I thought there was some weirdness with `os.Interrupt` on Windows? Maybe this was ages ago.

Reply:
SIGTERM seems to map to CTRL_CLOSE_EVENT, CTRL_LOGOFF_EVENT or CTRL_SHUTDOWN_EVENT so might as well add that. The docs seem to stats os.Interrupt will handle ctrl-c and break.

Reply:
Yes, ctrl-c works with `os.Interrupt` and `signal.Notify`, it's the comment in `os.Interrupt` I'm thinking of. "On Windows, sending os.Interrupt to a process with os.Process.Signal is not implemented; it will return an error instead of sending a signal."

Reply:
Changed to Interrupt and added SIGTERM also